File: backend/videosdb/ipfs.py
import os
import logging
import shutil
import sys
import tempfile
import ipfshttpclient
import videosdb.settings as settings
import socket
from google.oauth2 import service_account
from videosdb.youtube_api import YoutubeDL, parse_youtube_id
#from videosdb.models import Video
from google.cloud import firestore
from google.cloud import dns

logger = logging.getLogger(__name__)

BASE_DIR = os.path.dirname(sys.modules[__name__].__file__)


class DNS:
    def __init__(self, dns_zone):
        self.dns_zone = dns_zone
        creds_json_path = os.path.join(BASE_DIR, "creds-worpdress.json")
        self.client = dns.Client(
            credentials=service_account.Credentials.from_service_account_file(creds_json_path))

    def _update_record(self, record_name, record_type, ttl, new_value):

        if not self.dns_zone:
            return

        zone = self.client.zone(self.dns_zone)
        records = zone.list_resource_record_sets()

        # init transaction
        changes = zone.changes()
        # delete old
        for record in records:
            if record.name == record_name + "." and record.record_type == record_type:
                changes.delete_record_set(record)
        # add new
        record = zone.resource_record_set(
            record_name + ".", record_type, ttl, [new_value, ])
        changes.add_record_set(record)
        # finish transaction
        changes.create()

    def update_dnslink(self, record_name, new_root_hash):
        self._update_record(record_name, "TXT", 300,
                            "dnslink=/ipfs/" + new_root_hash)

    def update_ip(self, record_name, new_ip):
        self._update_record(record_name, "A", 300, new_ip)


class IPFS:
    def __init__(self, files_root="/videos"):
        creds_json_path = os.path.join(BASE_DIR, "creds.json")
        self.db = firestore.Client(
            credentials=service_account.Credentials.from_service_account_file(creds_json_path))
        self.host = socket.gethostbyname(settings.IPFS_HOST)
        self.port = settings.IPFS_PORT
        self.files_root = files_root
        self.dnslink_update_pending = False
        self.api = ipfshttpclient.connect(
            "/ip4/%s/tcp/%s/http" % (self.host, self.port), session=True)
        self.api.files.mkdir(files_root, parents=True)

    def add_file(self, filename, add_to_dir=True, **kwargs):
        result = self.api.add(
            filename, pin=True, **kwargs)
        ipfs_hash = result[0]["Hash"]
        assert ipfs_hash

        if add_to_dir:
            self.add_to_dir(filename, ipfs_hash)

        return ipfs_hash

    def add_to_dir(self, filename, _hash):
        from ipfshttpclient.exceptions import StatusError
        src = "/ipfs/" + _hash
        dst = self.files_root + "/" + os.path.basename(filename)
        try:
            self.api.files.rm(dst)
        except StatusError:
            pass
        self.api.files.cp(src, dst)
        self.dnslink_update_pending = True

    def get_file(self, ipfs_hash):
        self.api.get(ipfs_hash)
        files = os.listdir(".")
        filename = max(files, key=os.path.getctime)
        return filename

    def update_dnslink(self, force=False):
        if not self.dnslink_update_pending and not force:
            return

        root_hash = self.api.files.stat(self.files_root)["Hash"]
        dns = DNS(settings.VIDEOSDB_DNSZONE)
        dns.update_dnslink("videos." +
                           settings.VIDEOSDB_DOMAIN, root_hash)
        self.dnslink_update_pending = False

    def _files_in_ipfs_dict(self):
        # 'Entries': [
        #     {'Size': 0, 'Hash': '', 'Name': 'Software', 'Type': 0}
        # ]
        files = self.api.files.ls(self.files_root, opts=dict(long=True))
        files_in_ipfs = {}
        if files["Entries"]:
            for file in files["Entries"]:
                if file["Name"].lower().endswith(".mp4"):
                    youtube_id = parse_youtube_id(file["Name"])
                    if not youtube_id or youtube_id in files_in_ipfs:
                        raise Exception()
                    files_in_ipfs[youtube_id] = file
        return files_in_ipfs

    @staticmethod
    def _files_in_disk_dict(dir):
        files_in_disk = {}
        for file in os.listdir(dir):
            youtube_id = parse_youtube_id(file)
            if file.endswith(".part"):
                continue
            if not youtube_id or youtube_id in files_in_disk:
                raise Exception()
            files_in_disk[youtube_id] = file
        return files_in_disk

    def download_and_register_folder(self, overwrite_hashes=False):
        yt_dl = YoutubeDL()

        videos_dir = os.path.abspath(settings.VIDEO_FILES_DIR)
        if not os.path.exists(videos_dir):
            os.mkdir(videos_dir)

        files_in_ipfs_by_id = self._files_in_ipfs_dict()
        files_in_disk_by_id = self._files_in_disk_dict(videos_dir)

        videos_in_db_ids = self.db.collection("meta").document(
            "meta").get().to_dict()["videoIds"]

        for video_id in videos_in_db_ids:
            video_ref = self.db.collection("videos").document(video_id)

            if not video_id in files_in_disk_by_id:
                logger.debug("Downloading " + video_id)
                with tempfile.TemporaryDirectory(dir=videos_dir) as tmpdir:
                    os.chdir(tmpdir)
                    try:
                        filename = yt_dl.download_video(
                            video_id)
                    except YoutubeDL.UnavailableError as e:
                        continue
                    video_ref.set({
                        "videosdb": {
                            "filename": filename
                        }
                    }, merge=True)
                    try:
                        shutil.move(filename, videos_dir)
                    except OSError as e:
                        logger.exception(e)
                        continue

            file = files_in_disk_by_id.get(video_id)
            # if file and file != video.filename:
            #     video.filename = file

            if video_id in files_in_ipfs_by_id:
                file = files_in_ipfs_by_id[video_id]

                logger.debug("Already in IPFS:  " + str(file))
                video_ref.set({
                    "videosdb": {
                        "filename": file["Name"],
                        "ipfs_hash": file["Hash"]
                    }
                }, merge=True)
                continue

            # adding to IPFS:
            video_doc = video_ref.get()
            video = video_doc.to_dict()
            logger.debug("Adding to IPFS: ID:%s, title: %s, Filename: %s, Hash: %s" %
                         (video_id, video["snippet"]["title"], video["videosdb"].get("filename", ""), video["videosdb"].get("ipfs_hash", "")))
            ipfs_hash = self.add_file(videos_dir + "/" +
                                      video["videosdb"]["filename"],
                                      wrap_with_directory=True,
                                      nocopy=True)
            video_ref.set({
                "videosdb": {
                    "ipfs_hash": ipfs_hash
                }
            }, merge=True)

        self.update_dnslink()

    # def _list_amule_shares(self):
    #     result = execute('amulecmd -P amule -c "show shared"')

File: backend_julia/src/ipfs.jl
module VideosDB.IPFS

using Logging
using JSON3
using Sockets
using Dates
using ..Settings: VIDEO_FILES_DIR, IPFS_HOST, IPFS_PORT, VIDEOSDB_DOMAIN, VIDEOSDB_DNSZONE
using ..YoutubeAPI: parse_youtube_id

export DNS, IPFS, parse_youtube_id

struct DNS
    dns_zone::String
end

function DNS(dns_zone::AbstractString)
    return DNS(dns_zone)
end

function _update_record(d::DNS, record_name::AbstractString, record_type::AbstractString, ttl::Int, new_value::AbstractString)
    if isempty(d.dns_zone)
        return nothing
    end
    # Placeholder: in production this should call Google DNS API. For tests we just log.
    @info("DNS update: $record_name $record_type -> $new_value")
    return true
end

function update_dnslink(d::DNS, record_name::AbstractString, new_root_hash::AbstractString)
    return _update_record(d, record_name, "TXT", 300, "dnslink=/ipfs/" * new_root_hash)
end

function update_ip(d::DNS, record_name::AbstractString, new_ip::AbstractString)
    return _update_record(d, record_name, "A", 300, new_ip)
end

mutable struct IPFS
    files_root::String
    files::Dict{String,String}
    dnslink_update_pending::Bool
    host::String
    port::Int
end

function IPFS(files_root::AbstractString = "/videos")
    host = IPFS_HOST
    port = IPFS_PORT
    ip = IPFS(files_root, Dict{String,String}(), false, host, port)
    return ip
end

# Add file: simulate an IPFS add and return a fake hash
function add_file(ip::IPFS, filename::AbstractString; add_to_dir::Bool=true; kwargs...)
    # simple fake hash: hex of timestamp + filename
    h = hex(Int(floor(time())))[end-12:end] * string(abs(hash(filename)))[end-6:end]
    if add_to_dir
        add_to_dir_func(ip, filename, h)
    end
    return h
end

function add_to_dir_func(ip::IPFS, filename::AbstractString, _hash::AbstractString)
    basename = splitpath(filename)[end]
    ip.files[basename] = _hash
    ip.dnslink_update_pending = true
    return true
end

function get_file(ip::IPFS, ipfs_hash::AbstractString)
    # In real client, this would fetch; here return a synthetic filename if present
    for (name,h) in ip.files
        if h == ipfs_hash
            return name
        end
    end
    return nothing
end

function update_dnslink(ip::IPFS; force::Bool=false)
    if !ip.dnslink_update_pending && !force
        return nothing
    end
    # compute a "root_hash" (fake)
    root_hash = join(values(ip.files), ",")
    dns = DNS(VIDEOSDB_DNSZONE)
    update_dnslink(dns, "videos." * VIDEOSDB_DOMAIN, root_hash)
    ip.dnslink_update_pending = false
    return root_hash
end

function _files_in_ipfs_dict(ip::IPFS)
    files_in_ipfs = Dict{String,Any}()
    for (name,h) in ip.files
        if endswith(lowercase(name), ".mp4")
            yid = parse_youtube_id(name)
            if yid === nothing || haskey(files_in_ipfs, yid)
                error("Invalid IPFS state or duplicate")
            end
            files_in_ipfs[yid] = Dict("Name"=>name, "Hash"=>h)
        end
    end
    return files_in_ipfs
end

function _files_in_disk_dict(dir::AbstractString)
    files_in_disk = Dict{String,String}()
    if !isdir(dir)
        return files_in_disk
    end
    for file in readdir(dir)
        if endswith(file, ".part")
            continue
        end
        yid = parse_youtube_id(file)
        if yid === nothing || haskey(files_in_disk, yid)
            error("Invalid disk state or duplicate")
        end
        files_in_disk[yid] = file
    end
    return files_in_disk
end

function download_and_register_folder(ip::IPFS; overwrite_hashes::Bool=false)
    # This is a simplified emulation: check VIDEO_FILES_DIR, compare, and "add" missing files
    videos_dir = abspath(get(ENV, "VIDEO_FILES_DIR", VIDEO_FILES_DIR))
    if !isdir(videos_dir)
        mkpath(videos_dir)
    end
    files_in_ipfs = _files_in_ipfs_dict(ip)
    files_in_disk = _files_in_disk_dict(videos_dir)
    # For the sake of example, add each file from disk to ipfs if missing
    for (yid, file) in files_in_disk
        if haskey(files_in_ipfs, yid)
            continue
        end
        path = joinpath(videos_dir, file)
        h = add_file(ip, path; add_to_dir=true)
        # In real client, update firestore doc; here just log
        @info("Added to IPFS: $file -> $h")
    end
    update_dnslink(ip)
    return true
end

end # module
